### ISSUE TITLE
Long compilation times

### ISSUE THREAD

**Issue by galexv** [ALPSCore/ALPSCore#212]
Code that involves accumulators takes a long time to compile. (This issue is long-known, just wanted to have it registered.) Lukas (@gamperl) wrote some code to help with this problem, so we can continue in the same direction.

**Comment by shinaoka** [comment]
I am now trying to build ALPSCore on a FX10 system at ISSP with a Fujitsu C++ compiler. An issue is that the compilation of alps::accumulators (not application codes) takes forever even with -O0.  I want to know bit more about the issue you mentioned.

**Comment by egull** [comment]
Oh, if accumulators has problems... how about archive? That's currently the one that takes the longest to compile. It's in the hdf5 module... same problem there?

**Comment by shinaoka** [comment]
Yeah, I have the same problem with archive.

**Comment by galexv** [comment]
Mario (@mskoenz) reports extremely long compilation times too (with GCC 5.2.1) on a machine that is used to compile tricky template metaprograms (could the slowdown be because of GTest?).

I am going to collect some statistics on the compilation times and memory requirements for each file in the project and see if i can find the culprit.

**Comment by egull** [comment]
Hiroshi had to give up FX10, so that part is closed. Re @mskoenz question: I don't think anybody ever tried GCC 5.2 for the library. That may be the culprit? Do we know more / if this happens with other compilers too?

**Comment by egull** [comment]
outdated with alea, wontfix

**Comment by krivenko** [comment]
> outdated with alea, wontfix

No, it is not.

Accumulators are still a part of the library, and so are their unit tests.
I highly doubt that ALEA will replace the accumulators in all dependent codes any time soon, which means I'll have to waste minutes and hours staring at make's output.

**Comment by egull** [comment]
Ok. Do you want to have a go at speeding it up? Alex doesn't have capacity for this and it hasn't been worked on for two years. How important is it for you?

**Comment by galexv** [comment]
Lukas Gamper's (@gamperl) approach was, if i understand correctly, to use PIMPL idiom: instantiate necessary templates explicitly and use pointers to them in the client code. This way all the lengthy compilation can be done only once. 

Also, there was a person (Ian George) recently interested with helping with ALPSCore: may this be a good project for him?

**Comment by mwallerb** [comment]
@krivenko  IMHO the main problem with the accumulator library is the deep dependency chain.  Thus each accumulator *instantiation* pulls in a convoluted inheritance hierarchy and some HDF5, MPI and whatnot on top of it.  This is what makes the tests very slow.

If you really want to have a go at it (which would be extremely appreciated!), I would suggest the following course of action:

  1. Create package `alps/alea/bridge` with the types: 
     - `NoBinningAccumlator`, 
     - `LogBinningAccumulator`, and
     - `FullBinningAccumulator`,
    specialized for `double` and `std::vector<double>` (others are unused) as thin wrappers over the respective alea classes.

 2. Move the simple unit tests to use these classes.  Implement *only* that functionality that they require.

 3. Retire alps::accumulators

I don't think this is *that* much work but it will cost a couple of days I think.

**Comment by egull** [comment]
OK. I'm stuck here. To make Igor happy we could:

1. Invest in fixing the accumulators. That's a non-trivial amount of work for a code that will not benefit anybody anymore, that's why this was closed.

2. Invest in a bridging layer (what @mwallerb suggests). That's also a lot of work, but it would allow us to get rid of accumulators soon, and old code can use the new functionality in a transparent way. That's probably the way to go, but should be done by @galexv or @mwallerb, because they know the code inside out. A lot of testing will be needed for this.

3. Mark the accumulators as deprecated, disable them by default but allow users to enable them (and their tests) where needed. Or enable them by default but allow Igor to disable them because he doesn't need them.

4. Do a quick fix that cuts down compilation time, if we can find one.

Any other ideas or suggestions? I'm tending towards (3) in the short time (default on, but Igor can switch them off), and (2) in the longer term... I'm very hesitant to make alea default at this point, it's not yet been tested enough.

**Comment by galexv** [comment]
If the old accumulators are actually not *used* InchwormHPX but just stand in the way, i think it is straightforward to implement selected compilation/building of ALPSCore components.

**Comment by galexv** [comment]
...and it may be  possible to decrease compilation time of accumulator tests: some of them use template based [type-parametrized Google Tests](https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#typed-tests). Preprocessor-based approach may compile faster.

**Comment by krivenko** [comment]
I've begun refactoring the accumulators and the hdf5 library.

https://github.com/krivenko/ALPSCore/tree/CLEANUP

There are a few reasons why things compile so slowly. Here is a list, in order of importance.

1. Translation units are huge. `libalps-hdf5.so` is made of one cpp file. `libalps-accumulators.so
` contains code from three large translation units. This is a problem not only because compilation time scales non-linearly with the size of the input, but also because `make -j` is effectively useless for those libraries. I split `archive.cpp` into 7 parts, and now `make -j4` builds `libalps-hdf5.so` two times faster than before.

2. Abuse of preprocessor macros for code generation. Those macros are enormous bundles of classes and functions copied and pasted many times with the only changing thing being one type. This task is what templates were invented for.

3. Metaprogramming insanity in the accumulators. I mean thing like member type/function presence detection and base class chaining. The same metaprogramming logic can be implemented in a much cleaner way with C++11 (or maybe even without it).

4. Use of Boost.MPL and Boost.TypeTraits, which are a serious hit to the compilation speed. 
These guys are purported to work with any compiler out there, no matter how broken it is. This is why their headers contain a metric ton of compile-time sanity checks and workarounds. I've almost completely removed both of them and switched to standard C++ metaprogramming devices from `<type_traits>`. Some bits of MPL are left in `include/alps/params/dict_types.hpp` because I didn't want to interfere with Alex's work.

**Comment by mwallerb** [comment]
@krivenko this is impressive work, congrats!

Also @egull, we should however still keep the alea bridge thing in the back of our heads, because the nice thing with that would be that alea gets a lot of unit tests in one go like that :-)

**Comment by krivenko** [comment]
> the nice thing with that would be that alea gets a lot of unit tests in one go like that :-)

How many of those tests would be compatible with the simple bridge? Most of them look complicated and bound to `accumulator_set`/`result_set` interfaces.

**Comment by krivenko** [comment]
@egull @mwallerb 
I've just realized that the old accumulators support arbitrary data types of observables. One can define `ALPS_ACCUMULATOR_USER_TYPES` to augment the list of the 6 standard data types. There is even an extensive unit test for this feature called `custom_type`.

The price to pay for this feature is accumulators being an almost entirely header-only library.
May I assume that `ALPS_ACCUMULATOR_USER_TYPES` is never used in practice?

**Comment by galexv** [comment]
> May I assume that ALPS_ACCUMULATOR_USER_TYPES is never used in practice?

@krivenko  Yes, i would say so.
It was a user request for matrix-like types, but i believe alea can handle the same problem differently. (@mwallerb is it correct?)

**Comment by krivenko** [comment]
Okay, so the question is about reverting the solution to #157...

**Comment by krivenko** [comment]
@egull What is your opinion on the user-defined types in accumulators? Is it okay to sacrifice this feature for faster compilation?

**Comment by egull** [comment]
Yes it is

**Comment by krivenko** [comment]
@galexv Dear Alex, could you please explain me what's the purpose of `/accumulators/include/alps/accumulators_.hpp`? This file seems to contain some preliminary effort at refactoring of the accumulators. Is it used somewhere?

**Comment by galexv** [comment]
Yes, this is an effort by Lukas (@gamperl) to refactor the accumulator code (as was mentioned earlier in https://github.com/ALPSCore/ALPSCore/issues/212#issuecomment-364620993). Unfortunately, we never had enough man-hours to continue this. I do not think it is actually used anywhere except a single test file `accumulators/test/mean_wrapped.cpp` and the (preliminary) implementation code `accumulators/src/accumulators.cpp`.

**Comment by krivenko** [comment]
Should be fixed by #458 and #464.
